### DISCUSSION TITLE
How to create detached variables?

### DISCUSSION THREAD

**Discussion by mplxrg** [EnzymeAD/Enzyme#1079] [Q&A]
In PyTorch, we can detach a tensor from the computation graph using` tensor.detach()`. I wondered if it is possible to achieve similar thing in Enzyme.

**Comment by tgymnich** [comment]
This is not quite the same as`detach()` but maybe it helps: When calling `enzyme_autodiff` you can mark variables as inactive with `enzyme_const`, which you can read more about in the [documentation](https://enzyme.mit.edu/getting_started/CallingConvention/#types).
[Compiler Explorer example](https://enzyme.gymni.ch/#g:!((g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:19,positionColumn:2,positionLineNumber:19,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'%23include+%3Cstdio.h%3E%0A%0Aint+enzyme_const%3B%0Aextern+double+__enzyme_autodiff(void*,+double...)%3B%0A%0Adouble+foo(double+x,+double+v)+%7B%0A++++double+y+%3D+x+*+x%3B%0A++++double+z+%3D+x+*+x+*+v%3B%0A++++double+r+%3D+y+%2B+z%3B%0A++++return+z%3B%0A%7D%0A%0Adouble+dfoo(double+x,+double+v)+%7B%0A++++return+__enzyme_autodiff((void*)foo,+x,+enzyme_const,+v)%3B%0A%7D%0Aint+main()+%7B%0A++++for(double+i%3D1%3B+i%3C5%3B+i%2B%2B)%0A++++++++printf(%22square(%25f)%3D%25f,+dsquare(%25f)%3D%25f%22,+i,+foo(i,3.0),+i,+dfoo(i,i%2B3.0))%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang12-enzyme-main,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'clang+12+(912a7f33da9)+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',m:69.67320261437908,n:'0',o:'',t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:9,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+clang+12+(912a7f33da9)+(Compiler+%231)',t:'0')),header:(),l:'4',m:30.326797385620917,n:'0',o:'',s:0,t:'0')),l:'3',n:'0',o:'',t:'0')),version:4)

**Comment by mplxrg** [comment]
@tgymnich Thank you so much for your answer! I'm trying to detach a intermediate variable inside a very complex function not a parameter of the function. I try to use the `enzyme_const` and write some extra code to achieve similar thing. But I still wondered if it is possible to mark intermediate variables as inactive inside a function body?

**Comment by wsmoses** [comment]
You can currently emulate much of this behavior by marking a function as inactive see [here](https://enzyme.gymni.ch/#g:!((g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'%23include+%3Cstdio.h%3E%0A%0Aint+enzyme_const%3B%0Aextern+double+__enzyme_autodiff(void*,+double...)%3B%0A%0A__attribute__((noinline))%0Adouble+identity(double+x)%3B%0A%0Aauto+__enzyme_inactivefn+%3D+identity%3B%0A%0Adouble+foo(double+x)+%7B%0A++++double+y+%3D+identity(x)%3B%0A++++return+x+%2B+y%3B%0A%7D%0A%0Adouble+dfoo(double+x)+%7B%0A++++return+__enzyme_autodiff((void*)foo,+x)%3B%0A%7D%0Aint+main()+%7B%0A++++for(double+i%3D1%3B+i%3C5%3B+i%2B%2B)%0A++++++++printf(%22square(%25f)%3D%25f,+dsquare(%25f)%3D%25f%22,+i,+foo(i),+i,+dfoo(i))%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang12-enzyme-main,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-S+-emit-llvm+-O2+-g0',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'clang+12+(51cd8528880)+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',m:69.67320261437908,n:'0',o:'',t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:9,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+clang+12+(51cd8528880)+(Compiler+%231)',t:'0')),header:(),l:'4',m:30.326797385620917,n:'0',o:'',s:0,t:'0')),l:'3',n:'0',o:'',t:'0')),version:4). This comes with the caveat that if the function is inlined, it will be marked as active again and contribute to the derivative.

It would be useful to understand precisely what your use case is and we can add an appropriate builtin.

For example, would it be useful to you to be able be able to mark a variable as inactive (meaning any uses of it are inactive)? [option A]

Or, is it more useful to you to be able to make a "copy" or "view" of a variable [doesn't actually copy anything] that doesn't impact the derivative. Any uses of the previous variable, however, will still impact the derivative.

Or perhaps the `__enzyme_inactivefn` marker on functions is sufficient for your use case.

**Comment by mplxrg** [comment]
Option B is exactly what I'm seeking. `__enzyme_inactivefn` seems to be a new feature in v0.0.17 which works pretty well for my use case. One thing to note here is that `__attribute__((noinline))` doesn't work for my compiler. The function still gets inlined. I opted to use `__attribute__((optnone))` to prevent the optimization.

**Comment by wsmoses** [comment]
Circling back on this, I'd be interested in making this an intrinsic directly (e.g. an `__enzyme_inactiveval` or something). Thoughts on name?

**Comment by mplxrg** [comment]
That would be great! `__enzyme_inactiveval` sounds good to me. Another name I could come up with is `__enzyme_detach()` which is similar to torch.detach().
